haystack/components/preprocessors/markdown_header_splitter.py

```diff
@@ -0,0 +1,391 @@
+
+import re
+from typing import Literal, Optional
+
+from haystack import Document, component, logging
+from haystack.components.preprocessors import DocumentSplitter
+
+logger = logging.getLogger(__name__)
+
+
+@component
+class MarkdownHeaderSplitter:
+    """
+    Split documents at Markdown headers, with optional secondary splitting and header level inference.
+
+    This component processes text documents by:
+    - Splitting them into chunks at Markdown headers (e.g., '#', '##', etc.), preserving header hierarchy as metadata.
+    - Optionally inferring and rewriting header levels for documents where header structure is ambiguous.
+    - Optionally applying a secondary split (by word, passage, period, or line) to each chunk.
+      This is done in haystack's DocumentSplitter.
+    - Preserving and propagating metadata such as parent headers, page numbers, and split IDs.
+    """
+
+    def __init__(
+        self,
+        *,
+        infer_header_levels: bool = False,
+        page_break_character: str = "\f",
+        secondary_split: Literal["none", "word", "passage", "period", "line"] = "none",
+        split_length: int = 200,
+        split_overlap: int = 0,
+        split_threshold: int = 0,
+    ):
+        """
+        Initialize the MarkdownHeaderSplitter.
+
+        :param infer_header_levels: If True, attempts to infer and rewrite header levels based on content structure.
+            Useful for documents where all headers use the same level. Defaults to False.
+        :param page_break_character: Character used to identify page breaks. Defaults to form feed ("\f").
+        :param secondary_split: Optional secondary split condition after header splitting.
+            Options are "none", "word", "passage", "period", "line". Defaults to "none".
+        :param split_length: The maximum number of units in each split when using secondary splitting. Defaults to 200.
+        :param split_overlap: The number of overlapping units for each split when using secondary splitting.
+            Defaults to 0.
+        :param split_threshold: The minimum number of units per split when using secondary splitting. Defaults to 0.
+        """
+        self.infer_header_levels = infer_header_levels
+        self.page_break_character = page_break_character
+        self.secondary_split = secondary_split
+        self.split_length = split_length
+        self.split_overlap = split_overlap
+        self.split_threshold = split_threshold
+
+        # initialize secondary_splitter only if needed
+        if self.secondary_split != "none":
+            self.secondary_splitter = DocumentSplitter(
+                split_by=self.secondary_split,
+                split_length=self.split_length,
+                split_overlap=self.split_overlap,
+                split_threshold=self.split_threshold,
+            )
+
+    def _infer_header_levels(self, text: str) -> str:
+        """
+        Infer and rewrite header levels in the markdown text.
```

Comment:
Maybe worth mentioning what style of markdown we use when doing this? I could be wrong here but it's not guaranteed that the header levels are using `#` right?